src/ast/query.rs

```diff
@@ -1642,6 +1645,62 @@ impl fmt::Display for OrderByExpr {
             Some(false) => write!(f, " NULLS LAST")?,
             None => (),
         }
+        if let Some(ref with_fill) = self.with_fill {
+            write!(f, " {}", with_fill)?
+        }
+        Ok(())
+    }
+}
+
+/// ClickHouse `WITH FILL` modifier for `ORDER BY` clause.
+#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "visitor", derive(Visit, VisitMut))]
+pub struct WithFill {
+    pub from: Option<Expr>,
+    pub to: Option<Expr>,
+    pub step: Option<Expr>,
+    pub interpolate: Vec<Interpolation>,
```

Comment:
I wonder if this part is correct? from their docs they say the following
> INTERPOLATE can be applied to columns not participating in ORDER BY WITH FILL

similarly looking at the grammar in the doc, interpolate seems to be a single clause applied to the order by expression rather than to the individual with fills?

Reply:
Using the example from the [docs](https://clickhouse.com/docs/en/sql-reference/statements/select/order-by#order-by-expr-with-fill-modifier), it isn't valid to say `ORDER BY n INTERPOLATE (inter AS inter + 1)`:

```
SELECT n, source, inter FROM (
   SELECT toFloat32(number % 10) AS n, 'original' AS source, number as inter
   FROM numbers(10) WHERE number % 3 = 1
) ORDER BY n INTERPOLATE (inter AS inter + 1);

Syntax error: failed at position 163 ('INTERPOLATE') (line 4, col 14):

SELECT n, source, inter FROM (
   SELECT toFloat32(number % 10) AS n, 'original' AS source, number as inter
   FROM numbers(10) WHERE number % 3 = 1
) ORDER BY n INTERPOLATE (inter AS inter + 1);

Expected one of: token, Dot, OR, AND, IS NOT DISTINCT FROM, IS NULL, IS NOT NULL, BETWEEN, NOT BETWEEN, LIKE, ILIKE, NOT LIKE, NOT ILIKE, REGEXP, IN, NOT IN, GLOBAL IN, GLOBAL NOT IN, MOD, DIV, alias, AS, DESCENDING, DESC, ASCENDING, ASC, NULLS, COLLATE, WITH FILL, Comma, LIMIT, OFFSET, FETCH, SETTINGS, UNION, EXCEPT, INTERSECT, INTO OUTFILE, FORMAT, end of query
```

The `INTERPOLATE` must be associated with the `WITH FILL` modifier:

```
SELECT
    n,
    source,
    inter
FROM
(
    SELECT
        toFloat32(number % 10) AS n,
        'original' AS source,
        number AS inter
    FROM numbers(10)
    WHERE (number % 3) = 1
)
ORDER BY n ASC WITH FILL
INTERPOLATE ( inter AS inter + 1 )

Query id: f8146c67-565f-4e19-8fe8-5dcfd9531add

   ┌─n─┬─source───┬─inter─┐
1. │ 1 │ original │     1 │
2. │ 2 │          │     2 │
3. │ 3 │          │     3 │
4. │ 4 │ original │     4 │
5. │ 5 │          │     5 │
6. │ 6 │          │     6 │
7. │ 7 │ original │     7 │
   └───┴──────────┴───────┘

7 rows in set. Elapsed: 0.001 sec.
```

The `FROM`, `TO`, and `STEP` options can be left out, but there needs to be a `WITH FILL`


Example playground available here: https://fiddle.clickhouse.com/96df3a22-5fa1-4c0e-8be6-36d963373425

Reply:
I've updated the structure for `INTERPOLATE` in [7022e42](https://github.com/sqlparser-rs/sqlparser-rs/pull/1330/commits/7022e4227fe00ae4914688ad91e355effd8a159b) to match.

Thanks for catching this

Reply:
Just to be explicit since our comments overlapped a bit:


> Ah I see, I think that makes sense! So it seems to only be a representation issue - interpolate can't be specified without `With Fill` but it still part of the order_by clause rather than a single `OrderByExpr` right or I might be misunderstanding still?
> 
> my understanding is that:
> 
> * its not valid to have multiple `INTERPOLATE` in an order by expression? e.g. this is invalid
> 
> ```sql
> ORDER BY a WITH FILL INTERPOLATE, b WITH FILL INTERPOLATE ...
> ```

That is correct. There can only be a single `INTERPOLATE` at the end of an `ORDER BY`, and multiple `WITH FILL`s can be added. Having at least a single `WITH FILL` is required for `INTERPOLATE` to work at all.

> * Then if I pass in the following to the parser:
> 
> ```sql
> ORDER BY a WITH FILL, b WITH FILL, c WITH FILL INTERPOLATE
> ```
> 
> * I get back a list of order by expressions
> 
> ```
> vec![
> 	OrderByExpr{a, WithFill{interpolate}},
> 	OrderByExpr{b, WithFill{interpolate}},
> 	OrderByExpr{c, WithFill{interpolate}},
> ]
> ```
> 
> where potentially only one of those entries in the list can actually contain _the_ interpolate.
> 
> does this describe the current behavior?


I added a test called `parse_select_order_by_with_fill_interpolate` which has:

```sql
ORDER BY
    fname ASC NULLS FIRST WITH FILL FROM 10 TO 20 STEP 2
    lname DESC NULLS LAST WITH FILL FROM 30 TO 40 STEP 3
    INTERPOLATE (col1 AS col1 + 1)
```

and parsing that will give back:

```rust
vec![
    OrderByExpr {
        expr: Expr::Identifier(Ident::new("fname")),
        asc: Some(true),
        nulls_first: Some(true),
        with_fill: Some(WithFill {
            from: Some(Expr::Value(number("10"))),
            to: Some(Expr::Value(number("20"))),
            step: Some(Expr::Value(number("2"))),
        }),
        interpolate: None,
    },
    OrderByExpr {
        expr: Expr::Identifier(Ident::new("lname")),
        asc: Some(false),
        nulls_first: Some(false),
        with_fill: Some(WithFill {
            from: Some(Expr::Value(number("30"))),
            to: Some(Expr::Value(number("40"))),
            step: Some(Expr::Value(number("3"))),
        }),
        interpolate: Some(InterpolationArg::Columns(vec![Interpolation {
            column: Expr::Identifier(Ident::new("col1")),
            formula: Some(Expr::BinaryOp {
                left: Box::new(Expr::Identifier(Ident::new("col1"))),
                op: BinaryOperator::Plus,
                right: Box::new(Expr::Value(number("1"))),
            }),
        }]))
    },
]
```